expression/collation.go

```diff
@@ -114,13 +114,13 @@ var (
 
 	// collationPriority is the priority when infer the result collation, the priority of collation a > b iff collationPriority[a] > collationPriority[b]
 	collationPriority = map[string]int{
-		charset.CollationASCII:   0,
```

Comment:
I think we need to rewrite the whole function. Before unicode_ci, the assumption in the comment holds. After introducing unicode_ci, it doesn't hold anymore.

Reply:
yes, `unicode_ci` and `general_ci`  is incompatible, my solution is make `unicode_ci` and `general_ci` has same priority. Same priority in `collationPriority ` means they are incompatible, and we can use [rule No.2](https://pingcap.com/docs-cn/stable/character-set-and-collation/#%E8%A1%A8%E8%BE%BE%E5%BC%8F%E4%B8%AD%E6%8E%92%E5%BA%8F%E8%A7%84%E5%88%99%E7%9A%84-coercibility-%E5%80%BC). I will rewrite `checkIllegalMixCollation` `deriveCoercibilityForScarlarFunc` and `DeriveCollationFromExprs`. This is first step to simplify code